### PR TITLE
fix Issue 20101 - BetterC: Template instantiation in CTFE only contex…

### PR DESCRIPTION
**Pull request by WalterBright**
…t should skip codegen

The code that determines what is "compile only" or not is a mess. Took me a while to figure out what knob to turn.

### CODE REVIEW

**Comment by WalterBright** [comment]
This enables more use of the GC in CTFE for @nogc code. How much more, I'm not sure. GC checking is already turned off for most CTFE use. It's just that for this particular case, the nogc betterC check was in the code gen.

File: compiler/src/dmd/traits.d [hunk 1]
```diff
@@ -1676,7 +1676,7 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
             uint errors = global.startGagging();
             Scope* sc2 = sc.push();
             sc2.tinst = null;
-            sc2.minst = null;
+            sc2.minst = null;   // this is why code for these are not emitted to object file
```
Comment (rikkimax) [inline compiler/src/dmd/traits.d:1679]: I'm not finding where Scope is declared via GH and I've forgotten where it's located, but wouldn't it be good to also document this behavior there as well?

Comment (WalterBright) [reply compiler/src/dmd/traits.d:1679]: Why what is emitted and what isn't took me a while to find out. It's more complex than just minst. This is the only case I found that set minst to null. It could use a refactor, but I thought I'd leave a breadcrumb for the next guy wondering why traits compiles doesn't emit code.

Comment (WalterBright) [reply compiler/src/dmd/traits.d:1679]: Essentially, too many people have worked on this aspect and the result is incoherence.

Comment (rikkimax) [reply compiler/src/dmd/traits.d:1679]: Indeed. Knowing what state the AST is in, is for sure an annoying beast!

**Comment by WalterBright** [comment]
That went better than I expected.

File: compiler/src/dmd/e2ir.d [hunk 2]
```diff
@@ -6296,7 +6296,7 @@ Lagain:
     if (r == RTLSYM.MEMSET8 ||
         r == RTLSYM.MEMSET16 ||
         r == RTLSYM.MEMSET32 ||
-	r == RTLSYM.MEMSET64)
```
Comment (thewilsonator) [inline compiler/src/dmd/e2ir.d:6299]: Unrelated changes

File: compiler/src/dmd/toobj.d [hunk 3]
```diff
@@ -281,7 +281,7 @@ void write_instance_pointers(Type type, Symbol *s, uint offset)
 
 void toObjFile(Dsymbol ds, bool multiobj)
 {
-    //printf("toObjFile(%s)\n", ds.toChars());
+    //printf("toObjFile(%s %s)\n", ds.kind(), ds.toChars());
```
Comment (thewilsonator) [inline compiler/src/dmd/toobj.d:284]: ditto

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/e2ir.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/e2ir.d:6299]: found when I ran detab over all the source files

File: compiler/src/dmd/toobj.d (hunk 3 above)
Comment (WalterBright) [reply compiler/src/dmd/toobj.d:284]: I generally leave in incidental improvements to debugging code.

**Comment by ibuclaw** [comment]
Probably should target stable
